File: data-structures/go/src/binaryheap/binary_heap.go
package binaryheap

import (
	"cmp"
	"errors"
)

var ErrEmptyHeap = errors.New("BinaryHeap: heap is empty")

type BinaryHeap[T cmp.Ordered] struct {
	data []T
}

func New[T cmp.Ordered]() *BinaryHeap[T] {
	return &BinaryHeap[T]{}
}

func FromSlice[T cmp.Ordered](arr []T) *BinaryHeap[T] {
	h := &BinaryHeap[T]{
		data: make([]T, len(arr)),
	}
	copy(h.data, arr)
	for i := len(h.data)/2 - 1; i >= 0; i-- {
		h.heapifyDown(i)
	}
	return h
}

func (h *BinaryHeap[T]) Push(value T) {
	h.data = append(h.data, value)
	h.heapifyUp(len(h.data) - 1)
}

func (h *BinaryHeap[T]) Pop() (T, error) {
	var zero T
	if len(h.data) == 0 {
		return zero, ErrEmptyHeap
	}
	min := h.data[0]
	last := len(h.data) - 1
	h.data[0] = h.data[last]
	h.data = h.data[:last]
	if len(h.data) > 0 {
		h.heapifyDown(0)
	}
	return min, nil
}

func (h *BinaryHeap[T]) Peek() (T, error) {
	var zero T
	if len(h.data) == 0 {
		return zero, ErrEmptyHeap
	}
	return h.data[0], nil
}

func (h *BinaryHeap[T]) Size() int {
	return len(h.data)
}

func (h *BinaryHeap[T]) IsEmpty() bool {
	return len(h.data) == 0
}

func (h *BinaryHeap[T]) Clear() {
	h.data = h.data[:0]
}

func (h *BinaryHeap[T]) Clone() *BinaryHeap[T] {
	clone := &BinaryHeap[T]{
		data: make([]T, len(h.data)),
	}
	copy(clone.data, h.data)
	return clone
}

func (h *BinaryHeap[T]) heapifyUp(i int) {
	for i > 0 {
		parent := (i - 1) / 2
		if h.data[i] >= h.data[parent] {
			break
		}
		h.data[i], h.data[parent] = h.data[parent], h.data[i]
		i = parent
	}
}

func (h *BinaryHeap[T]) heapifyDown(i int) {
	n := len(h.data)
	for {
		smallest := i
		left := 2*i + 1
		right := 2*i + 2

		if left < n && h.data[left] < h.data[smallest] {
			smallest = left
		}
		if right < n && h.data[right] < h.data[smallest] {
			smallest = right
		}
		if smallest == i {
			break
		}
		h.data[i], h.data[smallest] = h.data[smallest], h.data[i]
		i = smallest
	}
}

File: data-structures/python/src/binary_heap.py
from typing import TypeVar, Generic, List, Iterator

T = TypeVar('T')


class BinaryHeap(Generic[T]):
    def __init__(self) -> None:
        self._data: List[T] = []

    def push(self, value: T) -> None:
        self._data.append(value)
        self._sift_up(len(self._data) - 1)

    def pop(self) -> T:
        if not self._data:
            raise IndexError("pop from empty heap")
        if len(self._data) == 1:
            return self._data.pop()
        result = self._data[0]
        self._data[0] = self._data.pop()
        self._sift_down(0)
        return result

    def peek(self) -> T:
        if not self._data:
            raise IndexError("peek from empty heap")
        return self._data[0]

    def size(self) -> int:
        return len(self._data)

    def is_empty(self) -> bool:
        return len(self._data) == 0

    def clear(self) -> None:
        self._data.clear()

    def copy(self) -> 'BinaryHeap[T]':
        clone: BinaryHeap[T] = BinaryHeap()
        clone._data = self._data.copy()
        return clone

    @staticmethod
    def from_array(arr: List[T]) -> 'BinaryHeap[T]':
        """Build a heap from an array.

        Note: Creates a shallow copy of the input array.
        """
        heap: BinaryHeap[T] = BinaryHeap()
        heap._data = list(arr)
        for i in range(len(heap._data) // 2 - 1, -1, -1):
            heap._sift_down(i)
        return heap

    def _sift_up(self, index: int) -> None:
        while index > 0:
            parent = (index - 1) // 2
            if self._data[index] < self._data[parent]:
                self._data[index], self._data[parent] = self._data[parent], self._data[index]
                index = parent
            else:
                break

    def _sift_down(self, index: int) -> None:
        size = len(self._data)
        while True:
            smallest = index
            left = 2 * index + 1
            right = 2 * index + 2
            if left < size and self._data[left] < self._data[smallest]:
                smallest = left
            if right < size and self._data[right] < self._data[smallest]:
                smallest = right
            if smallest == index:
                break
            self._data[index], self._data[smallest] = self._data[smallest], self._data[index]
            index = smallest

    def __len__(self) -> int:
        return len(self._data)

    def __bool__(self) -> bool:
        return len(self._data) > 0

    def __repr__(self) -> str:
        return f"BinaryHeap({self._data})"

    def __str__(self) -> str:
        return f"BinaryHeap(size={len(self._data)})"

    def __iter__(self) -> Iterator[T]:
        heap_copy = self.copy()
        while not heap_copy.is_empty():
            yield heap_copy.pop()
